README.md
# Bibliography Activity

Bibliography Activity is a bibliography manager and genarator.  It uses the
Harvard style, and the bibliography generator is inspired by the School
Library Association of South Australia's Online Referencing Generator [1].

[1] The SLASA's Rederencing Generator is non-free SAAS for schools.  More
    information: http://www.slasa.asn.au/home.html


NEWS
5

* New icon
* Import from Browse - Turn bookmarks into bibliographies the new wizard view.
    View the web page on the left and the bibliography formatter on the right
    so that you don't need to go back and forth between browse activity.  Plus,
    Bibliography activity will automatically insert the webpage URL, title and
    date accessed into the correct bibliography fields.
* Uses the PopWindow module

4

* Add collaberation support - you can now work with a friend to add, edit
  and delete bibliography items

3

* Export as Write document (abiword format)
* Improved scrolling (scrollbars now visible, ScrollingDetector)
* Use bundle launcher to open offer to open exported files
* Use sugar style palettes instaed of GtkMenu palettes
* Icon improvements
* Small visual improvements

2

* Fixed typo thanks to Quozl

1

* First release


bib_types.py
# Copyright 2014 Sam Parkinson
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import logging
from gettext import gettext as _

ALL_TYPES = {}
ALL_TYPE_NAMES = []
WEB_TYPES = []


class BibType(object):

    def __init__(self, type, name, items, format_func,
                 web_title=None, web_uri=None):
        self.type = type
        self.name = name
        self.format = format_func
        self.web_title_index = web_title
        self.web_uri_index = web_uri

        items = [tuple(item.strip().split(':')) for item in items.split('|')]
        # Fix issue with URLs having a colon
        self.items = [(item[0], ':'.join(item[1:])) for item in items]

        ALL_TYPES[self.name] = self
        ALL_TYPE_NAMES.append(self.name)

def basic_format(format_string):
    def closure(values):
        return format_string.format(*values)
    return closure

# NOTE: To use the and symbol (`&`), you must type `&amp;`
#       You also need to escape charecters like less than (`<` becomes `&lt;`)
#       and the greater than (`>` becomes `&gt;`)
#
# Format Information:
# The pipe charecter separates a tuple of the bibliography entry and
# placeholder/example.  The tuple is separated by a colon.
# DO NOT TRANSLATE words with an asterisk (`*`) at the start
#
# TODO:
# eBooks from Databases... does anybody have examples?
# Books with more than 2 authors
# CD recordings

BibType('Book', _('Book'),
        _('Last Name:Shoup | First Name Initial:K | Year of Publication:2008 |'
          'Title:Reuse your refuse | Publisher:Wiley |'
          'Place of Publication:Hoboken, N.J'),
        basic_format('{}, {} {}, <i>{}</i>, {}, {}'))
BibType('Book with 2 Authors', _('Book with 2 Authors'),
        _('Author 1 Last Name:Fiell | Author 1 First Name Initial:C |'
          'Author 2 Last Name:Fiell | Author 2 First Name Initial:P |'
          'Year of Publication:2005 | Title:Graphic design now |'
          'Publisher:Taschen | Place of Publication:London'),
        basic_format('{}, {} &amp; {}, {} {}, <i>{}</i>, {}, {}'))
BibType('Book without Author', _('Book without Author'),
        _('Title:Rome | Year of Publication:2008 |'
          'Publisher:Dorling Kindersley | Place of Publication:London'),
        basic_format('<i>{}</i> {}, {}, {}'))
BibType('Book with Editor', _('Book with Editor'),
        _('Editor Last Name:West | Editor First Name Initial:S |'
          'Year of Publication:2005 | Title:Guide to art |'
          'Publisher:Bloomsbury | Place of Publication:London'),
        basic_format('{}, {} (ed.) {}, <i>{}</i>, {}, {}'))


def ebook_format(without_edition, with_edition, index):
    def closure(values):
        if values[index].strip():
            return with_edition.format(*values)
        else:
            del values[index]
            return without_edition.format(*values)
    return closure

BibType('eBook', _('eBook'),
        _('Last Name:Sachar | First Name Initial:L |'
          'Year of Publication:2010 | Edition (if applicable): | Title:Holes |'
          'Publisher:Bloomsbury Publishing | Place of Publication:London |'
          'Accessed:*datenow |'
          'URL:http://books.google.com.au/books?id=vlJw20OoYqUC'),
        ebook_format('{}, {} {}, <i>{}</i>, {}, {}, accessed {}, &lt;{}&gt;',
            '{}, {} {}, {} edn, <i>{}</i>, {}, {}, accessed {}, &lt;{}&gt;',
            3))
BibType('eBook with 2 Authors', _('eBook with 2 Authors'),
        _('Author 1 Last Name:Sharpley | Author 1 First Name Initial:R |'
          'Author 2 Last Name:Telfer | Author 2 First Name Initial:D |'
          'Year of Publication:2002 | Edition (if applicable): | '
          'Title:Tourism and Development: Concepts and Issues |'
          'Publisher:Channel View Publications |'
          'Place of Publication:Bristol | Accessed:*datenow |'
          'URL:https://books.google.com.au/books?id=Wvo1sIjZH3UC'),
        ebook_format('{}, {} &amp; {}, {} {}, <i>{}</i>, {}, {}, '
                     'accessed {}, &lt;{}&gt;',
                     '{}, {} &amp; {}, {} {}, {} edn, <i>{}</i>, {}, {}, '
                     'accessed {}, &lt;{}&gt;',
                     5))
BibType('eBook without Author', _('eBook without Author'),
        _('Title: You\'ve got what? | Year of Publication:2009 |'
          'Edition (if applicable):4th |'
          'Publisher:Communicable Disease Control Branch, Department of Health'
          '| Place of Publication:Adelaide | Accessed:*datenow |'
          'URL:http://www.publications.health.sa.gov.au/cgi/viewcontent.cgi'
          '?article=1029'),
        ebook_format('<i>{}</i> {}, {}, {}, accessed {}, &lt;{}&gt;',
            '<i>{}</i> {}, {} edn, {}, {}, accessed {}, &lt;{}&gt;',
            2))

BibType('Electronic Encyclopedia', _('Electronic Encyclopedia'),
        _('Title of Article:Earthquake | Year of Publication:2013 |'
          'Title of Encyclopedia:Encyclopaedia Britannica |'
          'Accessed:*datenow | URL:http://www.school.eb.com.au/all/comptons/'
          'article-9274104?query=earthquake'),
        basic_format('\'{}\' {}, in <i>{}</i>, accessed {}, &lt;{}&gt;'))
BibType('Printed Encyclopedia with Author',
        _('Printed Encyclopedia with Author'),
        _('Last Name:Pettus | First Name Initial:A M |'
          'Year of Publication:1998 | Title of Article:Edward Jenne |'
          'Title of Encyclopedia:Biographical encyclopedia of scientists |'
          'Publisher:Marshall Cavendish |'
          'Place of Publication:Tarrytown, N.Y. | Volume Number:3 |'
          'Starting Page:691 | Finishing Page:693'),
        basic_format('{}, {} {}, \'{}\' in <i>{}</i>, {}, {} '
                     'vol. {}, pp. {}-{}'))
BibType('Printed Encyclopedia without Author',
        _('Printed Encyclopedia without Author'),
        _('Title of Article:Germany | Year of Publication:2008 |'
          'Title of Encyclopedia:The World Book | Publisher:World Book |'
          'Place of Publication:Chicago | Volume Number:8 |'
          'Starting Page:146 | Finishing Page:172'),
        basic_format('\'{}\' {} in <i>{}</i>, {}, {} vol. {}, pp. {}-{}'))

def vid_format(string, volumei, issuei, datei):
    def closure(values):
        volume = ' vol. {},'.format(values[volumei]) \
            if values[volumei].strip() else ''
        issue = ' no. {},'.format(values[issuei]) \
            if values[issuei].strip() else ''
        date = ' {},'.format(values[datei]) if values[datei].strip() else ''
        fvalues = values[:volumei] + [volume, issue, date] + values[datei + 1:]
        return string.format(*fvalues)
    return closure

BibType('Magazine or Journal Article with Author',
        _('Magazine or Journal Article with Author'),
        _('Last Name:Carter | First Name Initial:R |'
          'Year of Publication:2014 |'
          'Tite of Article:Take Control of Your Dreams |'
          'Title of Magazine:BBC Focus | Volume (if applicable): |'
          'Issue (if applicable):271 |'
          'Date of Issue (if applicable):August | Starting Page:37 |'
          'Finishing Page:43'),
        vid_format('{}, {} {} \'{}\', <i>{}</i>,{}{}{} pp. {}-{}', 5, 6, 7))
BibType('Magazine or Journal Article without Author',
        _('Magazine or Journal Article without Author'),
        _('Tite of Article:Appliances of Science | Year of Publication:2014 |'
          'Title of Magazine:BBC Focus | Volume (if applicable): |'
          'Issue (if applicable):271 |'
          'Date of Issue (if applicable):August | Starting Page:87 |'
          'Finishing Page:87'),
        vid_format('{}, {} {} \'{}\', <i>{}</i>,{}{}{} pp. {}-{}', 3, 4, 5))

BibType('Online Magazine or Journal Article with Author',
        _('Online Magazine or Journal Article with Author'),
        _('Last Name:Keneley | First Name Initial:M |'
          'Year of Publication:2004 | Tite of Article:The dying town syndrome:'
          'a survey of urban development in the Western District of Victoria'
          '1830 - 1930 | Title of Magazine:Electronic Journal of Australian'
          'and New Zealand History | Volume (if applicable): |'
          'Issue (if applicable): |'
          'Date of Issue (if applicable):19 February | Accessed:*datenow |'
          'URL:ttp://www.jcu.edu.au/aff/history/articles/keneley3.htm'),
        vid_format('{}, {} {} \'{}\', <i>{}</i>,{}{}{}'
                   ' accessed {}, &lt;{}&gt;', 5, 6, 7))
BibType('Online Magazine or Journal Article without Author',
        _('Online Magazine or Journal Article without Author'),
        _('Tite of Article:Logging off? | Year of Publication:2010 |'
          'Title of Magazine:New Internationalist | Volume (if applicable): |'
          'Issue (if applicable):432 | Date of Issue (if applicable):May |'
          'Accessed:*datenow | URL:http://www.newint.org/columns/currents/2010'
          '/05/01/illegal-logging-madagascar'),
        vid_format('\'{}\', {} <i>{}</i>,{}{}{} pp. {}-{},'
                   ' accessed {}, &lt;{}&gt;', 3, 4, 5))

def page_format(string, starti, endi):
    def closure(values):
        page = 'pp. {}-{}'.format(values[starti], values[endi])
        if values[starti].strip() == values[endi].strip() \
           or not values[endi].strip():
            page = 'p. {}'.format(values[starti])

        values[starti] = page
        del values[endi]
        return string.format(*values)
    return closure

BibType('Newsaper Article with Author',
        _('Newsaper Article with Author'),
        _('Last Name:Bourke | First Name Initial:L |'
          'Year of Publication:2014 |'
          'Title of Article:New push to hit online buys with GST |'
          'Title of Newspaper:The Canberra Times |'
          'Date of Issue:27 December | Starting Page:1 |'
          'Finishing Page:'),
        page_format('{}, {} {} \'{}\', <i>{}</i>, {}, {}', 6, 7))
BibType('Newsaper Article without Author',
        _('Newsaper Article without Author'),
        _('Title of Article:Aspirin put to the test | Year of Publication:2005 |'
          'Title of Newspaper:Advertiser |'
          'Date of Issue:18 January | Starting Page:23 |'
          'Finishing Page:'),
        page_format('\'{}\', {}, <i>{}</i>, {}, {}', 4, 5))

BibType('Online Newsaper Article with Author',
        _('Online Newsaper Article with Author'),
        _('Last Name:Bourke | First Name Initial:L |'
          'Year of Publication:2014 | Title of Article:AirAsia QZ8501: '
          'Australia joins search for missing AirAsia flight |'
          'Title of Newspaper:The Canberra Times |'
          'Date of Issue:29 December | Accessed:*datenow |'
          'URL:http://www.canberratimes.com.au/federal-politics/political-news'
          '/airasia-qz8501-australia-joins-search-for-missing-airasia-flight-'
          '20141229-12exkr.html'),
        basic_format('{}, {} {} \'{}\', <i>{}</i>, {},'
                     ' accessed {}, &lt;{}&gt;'))
BibType('Online Article without Author',
        _('Online Article without Author'),
        _('Title of Article:Google street view broke privacy law |'
          'Year of Publication:2010 |Title of Newspaper:Advertiser |'
          'Date of Issue:9 July | Accessed:*datenow |'
          'URL:http://www.adelaidenow.com.au/google-street-view-broke-'
          'privacy-law/story-e6frea8c-1225890011209'),
        basic_format('\'{}\', {}, <i>{}</i>, {}, accessed {}, &lt;{}&gt;'))

def license_format(string, index=-1):
    def closure(values):
        license = ', License: &lt;{}&gt;'.format(values[index]) \
                  if values[index].strip() else ''
        values[index] = license
        return string.format(*values)
    return closure

WEB_TYPES.extend([ \
    BibType('Image with Creator (Real Name)',
            _('Image with Creator (Real Name)'),
            _('Last Name:Ganguly | First Name Initial:B | Year Created:2010 |'
              'Title or Description:Chicken Egg without Eggshell | Format:Photo |'
              'Sponsor or Orginisation:Wikimedia Commons | Accessed:*datenow |'
              'URL:http://commons.wikimedia.org/wiki/File:Chicken_Egg_without_'
              'Eggshell_5859.jpg | License URL (if available):http://commons.'
              'wikimedia.org/wiki/'
              'Commons:GNU_Free_Documentation_License,_version_1.2'),
            license_format('{}, {} {}, <i>{}</i>, {}, {},'
                           ' accessed {}, &lt;{}&gt;{}'),
            web_title=3, web_uri=7),
    BibType('Image with Creator (Screen Name)',
            _('Image with Creator (Screen Name)'),
            _('Screen Name or User Name:Dschwen | Year Created:2009 |'
              'Title or Description:Looking north from Chicago \'L\' station |'
              'Format:Photo | Sponsor or Orginisation:Wikimedia Commons |'
              'Accessed:*datenow | URL:http://commons.wikimedia.org/wiki/File:'
              'CTA_Night.jpg | License URL (if available):http://creativecommons.'
              'org/licenses/by-sa/4.0/deed.en'),
            license_format('{} {}, <i>{}</i>, {}, {}, accessed {}, &lt;{}&gt;{}'),
            web_title=2, web_uri=6),
    BibType('Image without Creator', _('Image without Creator'),
            _('Title or Description:OLPC XO Laptop with Screen Twisted |'
              'Year Created:n.d. | Format:Photo |'
              'Sponsor or Orginisation:One Laptop Per Child |'
              'Accessed:*datenow | URL:http://one.laptop.org/sites/default/files/'
              'hardware-left-side-view.png | License URL (if available):'),
            license_format('<i>{}</i> {}, {}, {}, accessed {}, &lt;{}&gt;{}'),
            web_title=0, web_uri=5),
    BibType('Website with Author', _('Website with Author'),
            _('Last Name:Lesinski | First Name Initial:K | Last Update:2014 |'
              'Title of Webpage:MozJPEG 3.0 |'
              'Sponsor or Orginisation:Performance Calendar | Accessed:*datenow |'
              'URL:http://calendar.perfplanet.com/2014/mozjpeg-3-0/'),
            basic_format('{}, {} {}, <i>{}</i>, {}, accessed {}, &lt;{}&gt;'),
            web_title=3, web_uri=6),
    BibType('Website by Organisation', _('Website by Organisation'),
            _('Name of Organisation:Sugar Labs | Last Update:2010 |'
              'Title of Webpage:Sugar Labs-learning software for children |'
              'Sponsor or Orginisation:Sugar Labs |'
              'Accessed:*datenow | URL:http://sugarlabs.org/'),
            basic_format('{} {}, <i>{}</i>, {}, accessed {}, &lt;{}&gt;'),
            web_title=2, web_uri=5),
    BibType('Website without Author', _('Website without Author'),
            _('Title of Webpage:Avocado Jackpot | Year Created:2014 |'
              'Sponsor or Orginisation:Reddit | Accessed:*datenow |'
              'URL:http://www.reddit.com/r/food/comments/2qnbpc/avocado_jackpot/'),
            basic_format('<i>{}</i> {}, {}, accessed {}, &lt;{}&gt;'),
            web_title=0, web_uri=4) \
])

def place_format(string, index):
    def closure(values):
        place = ', {}'.format(values[index]) if values[index].strip() else ''
        values[index] = place
        return string.format(*values)
    return closure

BibType('Film', _('Film'),
        _('Title:Toy Story 2 | Year Created:1999 | Format:DVD |'
          'Distributor:Buena Vista Home Entertainment |'
          'Place (if available): | Special Credits or Other Information: A '
          'Pixar Animation Studios Film'),
        place_format('<i>{}</i> {}, {}, {}{}. {}', 4))
BibType('Television Program (Single)', _('Television Program (Single)'),
        _('Title:Ten Bucks A Liter | Year of Broadcast:2013 | Format:iview |'
          'Television Channel:ABC | Place (if available): |'
          'Date of Broadcast:1 August'),
        place_format('<i>{}</i> {}, {}, {}{}, {}', 4))
BibType('Television Program (Part of Series)',
        _('Television Program (Part of Series)'),
        _('Episode Title:Radio Goodies | Year of Broadcast:1970 |'
          'Series Title: The Goodies | Format:DVD |'
          'Television Channel:BBC | Place (if available): |'
          'Date of Broadcast:20 December'),
        place_format('<i>{}</i> {}, {}, {}, {}{}, {}', 5))


add_button.py
# Copyright 2014 Sam Parkinson
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import GObject

from gettext import gettext as _

from sugar3.graphics.toolbutton import ToolButton
from sugar3.graphics.iconentry import IconEntry, ICON_ENTRY_PRIMARY
from sugar3.graphics import style


class AddToolButton(ToolButton):

    __gsignals__ = {
        'add-type': (GObject.SIGNAL_RUN_FIRST, None, (str,))
    }

    def __init__(self, types):
        ToolButton.__init__(self, 'list-add')
        self.set_tooltip(_('Add Bibliography Entry'))
        
        self.palette_invoker.props.toggle_palette = True
        self.palette_invoker.props.lock_palette = True
        self._p = self.get_palette()

        self._search_box = IconEntry()
        self._search_box.add_clear_button()
        self._search_box.set_icon_from_name(
            ICON_ENTRY_PRIMARY, 'system-search')
        self._search_box.show()

        types_store = Gtk.ListStore(str)
        for i in types:
            types_store.append([i])

        self._filter_model = types_store.filter_new()
        self._filter_model.set_visible_func(self.__model_filter_cb)
        self._search_box.connect('changed',
                                 lambda *args: self._filter_model.refilter())
        self._search_box.connect('activate', self.__search_box_activate_cb)

        treeview = Gtk.TreeView(self._filter_model)
        treeview.props.headers_visible = False
        try:
            treeview.props.activate_on_single_click = True
        except AttributeError:
            pass
        treeview.get_style_context().add_class('add-new-treeview')
        treeview.connect('row-activated', self.__row_clicked_cb)
        treeview.show()

        r = Gtk.CellRendererText()
        c = Gtk.TreeViewColumn(_('Type'), r, text=0)
        treeview.append_column(c)

        sw = Gtk.ScrolledWindow()
        sw.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.ALWAYS)
        sw.props.height_request = 300
        sw.add(treeview)
        sw.show()
        
        box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=6)
        box.add(self._search_box)
        box.add(sw)
        box.show()
        self._p.set_content(box)
        
    def __row_clicked_cb(self, treevieew, path, view_column):
        row = self._filter_model.get_iter(path)
        type_ = self._filter_model.get_value(row, 0)
        self.emit('add-type', type_)
        self._p.popdown()
        
    def __model_filter_cb(self, model, iter, data):
        return self._search_box.get_text().lower() in \
            model.get(iter, 0)[0].lower()

    def __search_box_activate_cb(self, entry):
        row = self._filter_model.get_iter(Gtk.TreePath.new_first())
        type_ = self._filter_model.get_value(row, 0)
        self.emit('add-type', type_)
        self._p.popdown()


activity.py
# Copyright 2014-2016 Sam Parkinson
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import os
import time
import json
import logging
from gettext import gettext as _

import dbus
from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import Pango

from sugar3.activity import activity
from sugar3.datastore import datastore
from sugar3.graphics.alert import Alert
from sugar3.graphics.toolbutton import ToolButton
from sugar3.graphics.toolbarbox import ToolbarBox
from sugar3.activity.widgets import ActivityToolbarButton
from sugar3.activity.widgets import StopButton
from sugar3.graphics import style
from sugar3.graphics.icon import Icon

from sugar3.datastore import datastore
from sugar3.graphics.objectchooser import ObjectChooser
from sugar3.graphics.objectchooser import FILTER_TYPE_ACTIVITY

try:
    from sugar3.activity.activity import get_bundle, launch_bundle
except ImportError:
    get_bundle = lambda **kwargs: None

try:
    from sugar3.presence.wrapper import CollabWrapper
    logging.error('USING SUGAR COLLAB WRAPPER!')
except ImportError:
    from textchannelwrapper import CollabWrapper

from add_button import AddToolButton
from add_window import EntryWindow
from browsewindow import BrowseImportWindow
from bib_types import ALL_TYPES, ALL_TYPE_NAMES
from main_list import MainList


class BibliographyActivity(activity.Activity):

    def __init__(self, handle):
        activity.Activity.__init__(self, handle)
        self._has_read_file = False
        self._collab = CollabWrapper(self)
        self._collab.message.connect(self.__message_cb)

        screen = Gdk.Screen.get_default()
        css_provider = Gtk.CssProvider.get_default()
        css_provider.load_from_path('style.css')
        context = Gtk.StyleContext()
        context.add_provider_for_screen(screen, css_provider,
                                        Gtk.STYLE_PROVIDER_PRIORITY_USER)

        toolbar_box = ToolbarBox()

        activity_button = ActivityToolbarButton(self)
        toolbar_box.toolbar.insert(activity_button, 0)
        activity_button.show()

        html = ToolButton('export-as-html')
        html.set_tooltip(_('Save as HTML'))
        html.connect('clicked', self.__export_as_html_cb)
        activity_button.props.page.insert(html, -1)
        html.show()

        abiword = ToolButton('export-as-abiword')
        abiword.set_tooltip(_('Save as a Write document'))
        abiword.connect('clicked', self.__export_as_abiword_cb)
        activity_button.props.page.insert(abiword, -1)
        abiword.show()

        add_button = AddToolButton(ALL_TYPE_NAMES)
        add_button.connect('add-type', self.__add_type_cb)
        toolbar_box.toolbar.insert(add_button, -1)
        add_button.show()

        browse = ToolButton('import-browse')
        browse.set_tooltip(_('Add Web Pages from Browse Entry'))
        browse.connect('clicked', self.__import_from_browse_cb)
        toolbar_box.toolbar.insert(browse, -1)
        browse.show()
   
        separator = Gtk.SeparatorToolItem()
        separator.props.draw = False
        separator.set_expand(True)
        toolbar_box.toolbar.insert(separator, -1)
        separator.show()

        stop_button = StopButton(self)
        toolbar_box.toolbar.insert(stop_button, -1)
        stop_button.show()

        self.set_toolbar_box(toolbar_box)
        toolbar_box.show()

        self._main_sw = Gtk.ScrolledWindow()
        self._main_sw.set_policy(Gtk.PolicyType.NEVER,
                                 Gtk.PolicyType.ALWAYS)
        self._main_sw.connect('key-press-event',
                              self.__key_press_event_cb)

        self._main_list = MainList(self._main_sw, self._collab)
        self._main_list.connect('edit-row', self.__edit_row_cb)
        self._main_list.connect('deleted-row', self.__deleted_row_cb)
        self._main_sw.add(self._main_list)
        self._main_list.show()

        self._empty_message = EmptyMessage()

        self.set_canvas(self._empty_message)
        self._empty_message.show()

        self._collab.setup()

    def add_item(self, text, type_, data):
        self._empty_message.hide()
        self.set_canvas(self._main_sw)
        self._main_sw.show()
        self._main_list.show()
        self._main_list.add(text, type_, data)

    def __message_cb(self, collab, buddy, msg):
        action = msg.get('action')
        if action is None:
            return

        args = msg.get('args')
        if action == 'add_item':
            self.add_item(*args)
        elif action == 'delete_row':
            self._main_list.delete(args)
        elif action == 'edit_item':
            self._main_list.edited_via_collab(msg.get('path'), args)
        else:
            logging.error('Got message that is weird %r', msg)

    def __add_type_cb(self, add_button, type_):
        window = EntryWindow(ALL_TYPES[type_], self)
        window.connect('save-item', self.__save_item_cb)
        window.show()
    
    def __save_item_cb(self, window, *args):
        self.add_item(*args)
        window.hide()
        self._collab.post(dict(
            action='add_item',
            args=args
        ))
        window.destroy()

    def __import_from_browse_cb(self, button):
        chooser = ObjectChooser(parent=self,
                                what_filter='org.laptop.WebActivity',
                                filter_type=FILTER_TYPE_ACTIVITY)
        result = chooser.run()

        if result == Gtk.ResponseType.ACCEPT:
            logging.debug('ObjectChooser: %r' % chooser.get_selected_object())
            jobject = chooser.get_selected_object()
            self._load_browse(jobject)

        chooser.destroy()
        del chooser

    def __try_again_cb(self, window, jobject):
        window.hide()
        window.destroy()
        self._load_browse(jobject)

    def _load_browse(self, jobject):
        if jobject and jobject.file_path:
            with open(jobject.file_path) as f:
                data = json.load(f)
            window = BrowseImportWindow(data, self, jobject)
            window.connect('save-item', self.__save_item_importer_cb)
            window.connect('try-again', self.__try_again_cb)
            window.show()

    def __save_item_importer_cb(self, window, *args):
        self.add_item(*args)
        self._collab.post(dict(
            action='add_item',
            args=args
        ))

    def __edit_row_cb(self, tree_view, type_, json_string):
        previous_values = json.loads(json_string)
        window = EntryWindow(ALL_TYPES[type_], self, previous_values)
        window.connect('save-item', tree_view.edited_row_cb)
        window.show()

    def __deleted_row_cb(self, tree_view, *row):
        if len(tree_view.get_model()) == 0:
            self._main_list.hide()
            self.set_canvas(self._empty_message)
            self._empty_message.show()

    def __key_press_event_cb(self, scrolled_window, event):
        keyname = Gdk.keyval_name(event.keyval)

        vadjustment = scrolled_window.props.vadjustment
        if keyname == 'Up':
            if vadjustment.props.value > vadjustment.props.lower:
                vadjustment.props.value -= vadjustment.props.step_increment
        elif keyname == 'Down':
            max_value = vadjustment.props.upper - vadjustment.props.page_size
            if vadjustment.props.value < max_value:
                vadjustment.props.value = min(
                    vadjustment.props.value + vadjustment.props.step_increment,
                    max_value)
        else:
            return False

    def __export_as_html_cb(self, button):
        jobject = datastore.create()
        jobject.metadata['title'] = \
            _('{} as HTML').format(self.metadata['title'])
        jobject.metadata['mime_type'] = 'text/html'
        preview = self.get_preview()
        if preview is not None:
            jobject.metadata['preview'] = dbus.ByteArray(preview)

        # write out the document contents in the requested format
        path = os.path.join(self.get_activity_root(),
                            'instance', str(time.time()))
        with open(path, 'w') as f:
            f.write('''<html>
                         <head>
                           <title>{title}</title>
                         </head>

                         <body>
                           <h1>{title}</h1>
                    '''.format(title=jobject.metadata['title']))
            for item in self._main_list.all():
                f.write('<p>{}</p>'.format(
                    item[self._main_list.COLUMN_TEXT]))
            f.write('''
                         </body>
                       </html>
                    ''')
        jobject.file_path = path

        datastore.write(jobject, transfer_ownership=True)
        self._journal_alert(jobject.object_id, _('Success'), _('Your'
                            ' Bibliography was saved to the journal as HTML'))
        jobject.destroy()
        del jobject

    def __export_as_abiword_cb(self, button):
        jobject = datastore.create()
        jobject.metadata['title'] = \
            _('{} as Write document').format(self.metadata['title'])
        jobject.metadata['mime_type'] = 'application/x-abiword'
        preview = self.get_preview()
        if preview is not None:
            jobject.metadata['preview'] = dbus.ByteArray(preview)

        path = os.path.join(self.get_activity_root(),
                            'instance', str(time.time()))
        with open(path, 'w') as f:
            f.write('<?xml version="1.0" encoding="UTF-8"?>\n<abiword>\n'
                    '<section>')
            entries = []
            for item in self._main_list.all():
                markup = item[self._main_list.COLUMN_TEXT]
                abiword = '<p><c>{}</c></p>'.format(markup) \
                    .replace('<b>', '<c props="font-weight:bold">') \
                    .replace('<i>', '<c props="font-style:italic;'
                             ' font-weight:normal">') \
                    .replace('</b>', '</c>').replace('</i>', '</c>')
                entries.append(abiword)
            f.write('\n<p><c></c></p>\n'.join(entries))
            f.write('</section>\n</abiword>')
        jobject.file_path = path

        datastore.write(jobject, transfer_ownership=True)
        self._journal_alert(jobject.object_id, _('Success'), _('Your'
                            ' Bibliography was saved to the journal as a Write'
                            ' document'))
        jobject.destroy()
        del jobject

    def _journal_alert(self, object_id, title, msg):
        alert = Alert()
        alert.props.title = title
        alert.props.msg = msg

        bundle = get_bundle(object_id=object_id)
        if bundle is not None:
            alert.add_button(Gtk.ResponseType.ACCEPT,
                             _('Open with {}').format(bundle.get_name()),
                             Icon(file=bundle.get_icon()))
        else:
            alert.add_button(Gtk.ResponseType.APPLY,
                             _('Show in Journal'),
                             Icon(icon_name='zoom-activity'))
        alert.add_button(Gtk.ResponseType.OK, _('Ok'),
                         Icon(icon_name='dialog-ok'))

        # Remove other alerts
        for alert in self._alerts:
            self.remove_alert(alert)

        self.add_alert(alert)
        alert.connect('response', self.__alert_response_cb, object_id)
        alert.show_all()

    def __alert_response_cb(self, alert, response_id, object_id):
        if response_id is Gtk.ResponseType.ACCEPT:
            launch_bundle(object_id=object_id)
        if response_id is Gtk.ResponseType.APPLY:
            activity.show_object_in_journal(object_id)
        self.remove_alert(alert)

    def write_file(self, file_path):
        if self._main_list is None:
            return  # WhataTerribleFailure
        data = self._main_list.all()
        with open(file_path, 'w') as f:
            json.dump(data, f)

        self.metadata['mime_type'] == 'application/json+bib'

    def get_data(self):
        return self._main_list.all()

    def read_file(self, file_path):
        # FIXME: Why does sugar call read_file so many times?
        if self._has_read_file:
            return
        self._has_read_file = True

        with open(file_path) as f:
            l = json.load(f)
        self.set_data(l)

    def set_data(self, l):
        if len(l) > 0:
            self._main_list.load_json(l)
            self._empty_message.hide()
            self.set_canvas(self._main_sw)
            self._main_sw.show()
            self._main_list.show()


class EmptyMessage(Gtk.EventBox):

    def __init__(self):
        Gtk.EventBox.__init__(self)
        self.modify_bg(Gtk.StateType.NORMAL,
                       style.COLOR_WHITE.get_gdk_color())

        alignment = Gtk.Alignment.new(0.5, 0.5, 0.1, 0.1)
        self.add(alignment)

        box = Gtk.VBox()
        alignment.add(box)

        icon = Icon(pixel_size=style.LARGE_ICON_SIZE,
                    icon_name='view-list',
                    stroke_color=style.COLOR_BUTTON_GREY.get_svg(),
                    fill_color=style.COLOR_BUTTON_GREY.get_svg())
        box.pack_start(icon, expand=True, fill=False, padding=0)

        label = Gtk.Label()
        color = style.COLOR_BUTTON_GREY.get_html()
        label.set_markup('<span weight="bold" color="%s">%s</span>' % (
            color, _('No Bibliography Entries')))
        box.pack_start(label, expand=True, fill=False, padding=0)

        self.show_all()
